src/openms/source/ANALYSIS/ID/IDMapper.cpp

```diff
@@ -190,253 +200,176 @@ namespace OpenMS
-              il.push_back(i);
-              fmap[fm_index].setMetaValue("spectrum_index", il);
-            }
-            ++assigned[i];
-          }
-
-          if (was_added) break;
-
-        } // m/z values to check
-      } // features
+        empty_protein_id.setIdentifier(map.getProteinIdentifications()[0].getIdentifier());
+      } else if (!map.getUnassignedPeptideIdentifications().empty())
+      {
+        empty_protein_id.setIdentifier(map.getUnassignedPeptideIdentifications()[0].getIdentifier());
+      } else
+      {
+        // No search run identifier given so we create a new one
+        empty_protein_id.setIdentifier("UNKNOWN_SEARCH_RUN_IDENTIFIER");
+        map.getProteinIdentifications().push_back(empty_protein_id);
+      }
     }
-  }
-
-  void IDMapper::annotate(ConsensusMap& cmap, const MSExperiment<>& pmap, bool measure_from_subelements, bool annotate_with_subelements)
-  {
-    //keep track of assigned/unassigned precursors
-    std::map<Size, Size> assigned;
 
-    // store which precursor fit which feature (and avoid double entries)
-    // consensusMap -> {spectrum_index}
-    std::vector<std::set<size_t> > mapping(pmap.size());
-
-    DoubleList mz_values;
-    IntList charges;
+    // for statistics:
+    Size spectrum_matches_none(0), spectrum_matches_single(0), spectrum_matches_multiple(0);
 
-    //iterate over the precursors
-    for (Size i = 0; i < pmap.size(); ++i)
+    // are there any mapped but unidentified precursors?
+    for (Size ui = 0; ui != unidentified.size(); ++ui)
     {
-      const vector<Precursor>& precursors = pmap[i].getPrecursors();
-
-      if (precursors.empty()) continue;
+      Size spectrum_index = unidentified[ui];
+      const MSSpectrum<Peak1D>& spectrum = spectra[spectrum_index];
+      const vector<Precursor>& precursors = spectrum.getPrecursors();
 
-      double rt_prec = pmap[i].getRT();
+      bool precursor_mapped(false);
 
-      //iterate over the features
-      for (Size cm_index = 0; cm_index < cmap.size(); ++cm_index)
+      // check if precursor has been identified
+      for (Size i_p = 0; i_p < precursors.size(); ++i_p)
       {
-        bool was_added = false; // was current precursor-m/z matched?!
+        // check by precursor mass and spectrum RT
+        double mz_p = precursors[i_p].getMZ();
+        int z_p = precursors[i_p].getCharge();
+        double rt_value = spectrum.getRT();
 
-        // iterate over m/z values of precursors
-        for (Size i_p = 0; i_p < precursors.size(); ++i_p)
+        PeptideIdentification precursor_empty_id;
+        precursor_empty_id.setRT(rt_value);
+        precursor_empty_id.setMZ(mz_p);
+        precursor_empty_id.setMetaValue("spectrum_index", spectrum_index);
+        if (!spectra[spectrum_index].getNativeID().empty())
         {
-          double mz_prec = precursors[i_p].getMZ();
+          precursor_empty_id.setMetaValue("spectrum_reference",  spectra[spectrum_index].getNativeID());
```

Comment:
I think we should warn in the file handlers. Warning in other classes might be possible but then you'll get potentially a lot of warnings.